### ISSUE TITLE
call executionEvents contains unpredictable time strings. 

### ISSUE THREAD

**Issue by ambrosejcarr** [broadinstitute/cromwell#2743]
[unpredictable_time_example.json.txt](https://github.com/broadinstitute/cromwell/files/1380844/unpredictable_time_example.json.txt)

In the attached metadata json file from a recent run (extension mangled with txt because github complains about json uploads) I'm seeing unpredictable datetime strings which, as a result, are difficult to parse. 

1. `2017-10-04T21:47:18.661720750Z` (the norm)
2. `2017-10-04T14:49:56.008-07:00` (`-07:00` suffix is also very common)
3.  `2017-10-04T21:54Z` (`calls: count.align_reads_split: executionEvents: "waiting for quota")

It doesn't seem to be restricted to start or end times.

This metadata file was obtained from the swagger REST metadata endpoint from a local version of cromwell 29. I verified it in browser and also downloaded it directly from the endpoint with requests.

request: `http://localhost:6361/api/workflows/v1/74e00c0a-ffb8-4e1a-94db-b0169ca7ed42/metadata`

**Comment by samanehsan** [comment]
I also noticed inconsistent datetime strings when querying workflows in the `https://cromwell.mint-dev.broadinstitute.org` cromwell. 

For example, the metadata returned from a request to `https://cromwell.mint-dev.broadinstitute.org/api/workflows/v2/0de7f0e7-51f9-44c5-ba39-0139d32362dc/metadata` includes a start time with no milliseconds: 
```
"end": "2017-09-15T01:50:58.022Z",
"start": "2017-09-15T01:50:58Z"
"submission": "2017-09-15T01:50:42.966Z"
```

**Comment by geoffjentry** [comment]
@samanehsan which version are you using? I think we've fixed that

**Comment by samanehsan** [comment]
@geoffjentry, which version has the fix? I believe the mint-dev cromwell is using version 29

**Comment by geoffjentry** [comment]
Oh disregard me. I just realized I was remembering @ambrosejcarr posting this issue in the first place

**Comment by kshakir** [comment]
With some discussion, we could maybe further canonicalize the format to make dates easier for users to read and compare. For example, we could always adjust offsets to UTC (aka `Z`), truncate milliseconds, etc.

Meanwhile, for those consuming the cromwell metadata in their programs, all of the above examples are valid [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) formats. Each example contains the date, time, and UTC-offset. Each also round-trips as is in Java/Scala.

```scala
import java.time._

val iso1 = OffsetDateTime.parse("2017-10-04T21:47:18.661720750Z")
val iso2 = OffsetDateTime.parse("2017-10-04T14:49:56.008-07:00")
val iso3 = OffsetDateTime.parse("2017-10-04T21:54Z")
val iso4 = OffsetDateTime.parse("2017-09-15T01:50:58.022Z")
val iso5 = OffsetDateTime.parse("2017-09-15T01:50:58Z")
val iso6 = OffsetDateTime.parse("2017-09-15T01:50:42.966Z")

println(s"iso1 = '$iso1'")
println(s"iso2 = '$iso2'")
println(s"iso3 = '$iso3'")
println(s"iso4 = '$iso4'")
println(s"iso5 = '$iso5'")
println(s"iso6 = '$iso6'")
```

```text
iso1 = '2017-10-04T21:47:18.661720750Z'
iso2 = '2017-10-04T14:49:56.008-07:00'
iso3 = '2017-10-04T21:54Z'
iso4 = '2017-09-15T01:50:58.022Z'
iso5 = '2017-09-15T01:50:58Z'
iso6 = '2017-09-15T01:50:42.966Z'
```

For those building JavaScript/EMCAScript front ends, the chrome console [parsed](https://www.ecma-international.org/ecma-262/5.1/#sec-15.9.1.15) all of the above into valid `Date` objects. I haven't tried it myself, but via google [this library](https://pypi.python.org/pypi/iso8601) looks promising for Python 2/3. I'm hopeful that other languages also support ISO 8601.

**Comment by ambrosejcarr** [comment]
@kshakir +1 for further harmonization. Thanks for the explanation about the different datetime strings. Do you happen to know why the different strings are occurring?

**Comment by kshakir** [comment]
I'm guessing that the different timezones / millisecond precisions are due to where the dates are generated. Dates are imported and/or stored a number of places, including ~JES~ PAPI, MySQL, etc.

I've frequently seen millisecond format differences when the formatter truncates the zero milliseconds. In the example above there appears to be a difference between [`start`](https://github.com/broadinstitute/cromwell/blob/2caec7d3fb24cef6e9e1fd1d3f89f6504f48dda8/engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowMetadataHelper.scala#L15) vs. [`end`](https://github.com/broadinstitute/cromwell/blob/2caec7d3fb24cef6e9e1fd1d3f89f6504f48dda8/engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/WorkflowMetadataHelper.scala#L20)/[`submission`](https://github.com/broadinstitute/cromwell/blob/2caec7d3fb24cef6e9e1fd1d3f89f6504f48dda8/engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreSubmitActor.scala#L111), ex: `"2017-09-15T01:50:58Z"` vs `"2017-09-15T01:50:58.000Z"`. Each is actually generated the same way using `OffsetDateTime.now.toString`.

For timezone differences, I'm a little less sure without more context. I believe most (all?) google dates come back as UTC. But it's possible one may run cromwell on a JVM using a [default offset](https://docs.oracle.com/javase/8/docs/api/java/util/TimeZone.html#getDefault--), producing some `OffsetDateTime.now` with `-07:00`. This is especially the case for dates stored in database `TIMESTAMP` columns, as we use the zone of _cromwell_ before [stripping the zone](https://github.com/broadinstitute/cromwell/blob/2caec7d3fb24cef6e9e1fd1d3f89f6504f48dda8/database/sql/src/main/scala/cromwell/database/sql/SqlConverters.scala#L9-L19) and sending the datetime over to the db.

A possible fix to this whole issue would be to feed all `OffsetDateTime`s through a function that, for example, does `.withNano(0).atZoneSameInstant(ZoneOffset.UTC)`, if that's what we want to do. Perhaps the ZoneOffset could be configurable, as I've heard cromwell users mentioning trouble doing the date math when staring at UTC times.

**Comment by samanehsan** [comment]
Hello! We are still seeing this issue on version `36-fde91e6` using Cromwell-as-a-Service. For example, in the workflow "d182afeb-33ae-45ac-8e78-ba9e0a7da7ab", the "SmartSeq2ZarrConversion" task has an end time with no seconds or milliseconds but the start time is the full date-time stamp:
```
"start": "2019-01-28T10:24:40.084Z" 
"end": "2019-01-28T10:47Z"
```

We noticed this issue in our pipeline that submits analysis results to the HCA, since the JSON schema that we follow to format the results requires that the timestamps are all in the full datetime format. 

We could handle this inconsistency in our pipeline, but having Cromwell return timestamps in a consistent format would be ideal because then the tools that consume this information would not have to implement individual work-arounds to standardize the dates.

**Comment by ruchim** [comment]
AC: Standardize things such that the date/time format is consistent across various keys of metadata. The date/time returned by Cromwell should be consistently UTC for now.

**Comment by kshakir** [comment]
Fixed going forward by #4959
